api/dps/server_integration_test.go

```diff
@@ -0,0 +1,1024 @@
+
+		storage := storage.New(codec)
+		reader := index.NewReader(db, storage)
+		writer := index.NewWriter(db, storage)
+
+		// Insert mock data in database.
+		require.NoError(t, writer.Header(mocks.GenericHeight, header))
+		require.NoError(t, writer.Close())
+
+		server := dps.NewServer(reader, codec)
+
+		req := &dps.GetHeaderRequest{
+			Height: mocks.GenericHeight,
+		}
+		resp, err := server.GetHeader(context.Background(), req)
+
+		require.NoError(t, err)
+		assert.Equal(t, mocks.GenericHeight, resp.Height)
+
+		wantData, err := codec.Marshal(header)
+		require.NoError(t, err)
+		assert.Equal(t, wantData, resp.Data)
+	})
+
+	t.Run("handles indexer failure on Header", func(t *testing.T) {
+		t.Parallel()
+
+		codec, err := zbor.NewCodec()
+		require.NoError(t, err)
+
+		db := helpers.InMemoryDB(t)
+
+		storage := storage.New(codec)
+		reader := index.NewReader(db, storage)
+
+		// No data is written in the database, so the index should fail to retrieve anything.
+
+		server := dps.NewServer(reader, codec)
+
+		req := &dps.GetHeaderRequest{
+			Height: mocks.GenericHeight,
+		}
+		_, err = server.GetHeader(context.Background(), req)
+
+		assert.Error(t, err)
+	})
+}
+
+func TestIntegrationServer_GetEvents(t *testing.T) {
+	events := mocks.GenericEvents(4)
+	height := mocks.GenericHeight
+
+	t.Run("nominal case without type specified", func(t *testing.T) {
+		t.Parallel()
+
+		codec, err := zbor.NewCodec()
+		require.NoError(t, err)
+
+		db := helpers.InMemoryDB(t)
+
+		storage := storage.New(codec)
+		reader := index.NewReader(db, storage)
+		writer := index.NewWriter(db, storage)
+
+		// Insert mock data in database.
+		require.NoError(t, writer.First(height))
+		require.NoError(t, writer.Last(height))
+		require.NoError(t, writer.Events(mocks.GenericHeight, events))
+		require.NoError(t, writer.Close())
+
+		server := dps.NewServer(reader, codec)
+
+		req := &dps.GetEventsRequest{
+			Height: height,
+		}
+		resp, err := server.GetEvents(context.Background(), req)
+
+		require.NoError(t, err)
+		assert.Equal(t, height, resp.Height)
+
+		var got []flow.Event
+		require.NoError(t, codec.Unmarshal(resp.Data, &got))
+		assert.ElementsMatch(t, events, got)
+	})
+
+	t.Run("nominal case with type specified", func(t *testing.T) {
+		t.Parallel()
+
+		codec, err := zbor.NewCodec()
+		require.NoError(t, err)
+
+		db := helpers.InMemoryDB(t)
+
+		storage := storage.New(codec)
+		reader := index.NewReader(db, storage)
+		writer := index.NewWriter(db, storage)
+
+		// Insert mock data in database.
+		require.NoError(t, writer.First(height))
+		require.NoError(t, writer.Last(height))
+		require.NoError(t, writer.Events(mocks.GenericHeight, events))
+		require.NoError(t, writer.Close())
+
+		server := dps.NewServer(reader, codec)
+
+		// Only request one of the types of the events. Since the generic events are
+		// of two types (to simulate deposit/withdrawal), we should get half of the
+		// events as a result of this request.
```

Comment:
That's a bit of implied behaviour we might want to clean up. Why not provide the event type to `GenericEvents`? Maybe as a second integer argument `n`, that corresponds to `mock.GenericEventType(n)`?

Then we could explicitly generate `deposits := mock.GenericEvents(2, 1)` and `withdrawals : = mock.GenericEvents(2, 2)`. The improvement is small, but I think it would remove some implicit behaviour that is not obvious.

Reply:
Will be done in https://github.com/optakt/flow-dps/issues/333